File: CS50X/C/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int check_sum(long n);
int add_digits(long number, long divisor, bool multiply_by_two);
int check_len(long n);
int check_first_digits(long n);

int main(void)
{
    long input = get_long("Number: ");
    int validity_identifier = check_sum(input);
    int number_of_digits = check_len(input);
    int first_digits = check_first_digits(input);

    if ((number_of_digits == 15) && (first_digits == 34 || first_digits == 37) &&
        (validity_identifier == 0))
    {
        printf("AMEX\n");
    }
    else if ((number_of_digits == 16) &&
             (first_digits == 51 || first_digits == 52 || first_digits == 53 ||
              first_digits == 54 || first_digits == 55) &&
             (validity_identifier == 0))
    {
        printf("MASTERCARD\n");
    }
    else if ((number_of_digits == 13 || number_of_digits == 16) && (first_digits / 10 == 4) &&
             (validity_identifier == 0))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

int check_sum(long n)
{
    int sum1 = add_digits(n, 10, true); // Calculate sum * 2 of every other digit, starting with the
                                        // number’s second-to-last digit.
    int sum2 = add_digits(n, 1, false); // Calculate sum of other digits.
    int last_digit = (sum1 + sum2) % 10; // Catch the last digit of sum1 + sum2.
    return last_digit;
}

int add_digits(long number, long divisor, bool multiply_by_two)
{
    int sum = 0;
    int multiplicator = 100;
    int target;

    do
    {
        target = (number / divisor) % 10;
        if (((target * 2) < 10) && multiply_by_two)
        {
            sum += target * 2;
        }
        else if (((target * 2) >= 10) && multiply_by_two) // If the target * 2 is higher than 9 we
                                                          // then add the digits of the target.
        {
            sum += ((target * 2) / 10) + ((target * 2) % 10);
        }
        else // Concerns digits that should not to be mulplied by 2 (multiply_by_two = false).
        {
            sum += target;
        }
        divisor *= multiplicator;
    }
    while ((number / divisor) > 0);
    return sum;
}

int check_len(long n)
{
    long divisor = 1;
    int number_of_digits = 0;
    long marker; // Create a marker to stop the loop when number_of_digits equals actual
                 // number of digits.
    do
    {
        marker = n / divisor;
        divisor *= 10;
        number_of_digits++;
    }
    while (marker > 10);
    return number_of_digits;
}

int check_first_digits(long n)
{
    long divisor = 1;
    long first_digits;
    do
    {
        first_digits = n / divisor;
        divisor *= 10;
    }
    while (first_digits > 100);
    return first_digits;
}

File: CS50X/Python/sentimental-credit/credit.py
def main():

    while True:
        try:
            number = int(input("Number: "))
            break
        except ValueError:
            continue

    validity_identifier = check_sum(number)
    number_of_digits = check_len(number)
    first_digits = check_first_digits(number)

    if (
        (number_of_digits == 15)
        and (first_digits == 34 or first_digits == 37)
        and (validity_identifier == 0)
    ):
        print("AMEX")

    elif (
        (number_of_digits == 16)
        and (
            first_digits == 51
            or first_digits == 52
            or first_digits == 53
            or first_digits == 54
            or first_digits == 55
        )
        and (validity_identifier == 0)
    ):

        print("MASTERCARD")

    elif (
        (number_of_digits == 13 or number_of_digits == 16)
        and (int(first_digits / 10) == 4)
        and (validity_identifier == 0)
    ):

        print("VISA")

    else:
        print("INVALID")


def check_sum(n):

    sum1 = add_digits(n, 10, True)
    sum2 = add_digits(n, 1, False)
    last_digit = int((sum1 + sum2) % 10)

    return last_digit


def add_digits(number, divisor, multiply=True):

    sum = 0
    multiplicator = 100

    while True:

        target = int((number / divisor) % 10)

        if int(number / divisor) > 0:

            if (target * 2) < 10 and multiply:

                sum += target * 2

            elif (target * 2) >= 10 and multiply:

                sum += int((target * 2) / 10) + int((target * 2) % 10)

            else:
                sum += target

            divisor *= multiplicator

        else:
            break

    return sum


def check_len(n):

    len = 0
    divisor = 1

    while True:
        marker = int(n / divisor)

        if marker > 0:
            divisor = divisor * 10
            len += 1

        else:
            break

    return len


def check_first_digits(n):

    divisor = 1
    while True:
        x = int(n / divisor)

        if x > 100:
            divisor *= 10

        elif x < 100:
            return x


main()
